### PR TITLE
WT-10025 Improve RTS logs

### PR DESCRIPTION
**Pull request by etienneptl**
Those changes bring consistency and add information related to the `page_del` structure when debugging RTS.

### CODE REVIEW

File: src/btree/bt_debug.c [hunk 1]
```diff
@@ -1638,6 +1639,12 @@ __debug_ref(WT_DBG *ds, WT_REF *ref)
     if (__wt_ref_addr_copy(session, ref, &addr) && !WT_TIME_AGGREGATE_IS_EMPTY(&addr.ta))
         WT_RET(ds->f(ds, ", %s, %s", __wt_time_aggregate_to_string(&addr.ta, time_string),
           __wt_addr_string(session, addr.addr, addr.size, ds->t1)));
+    if (ref->state == WT_REF_DELETED && ref->page_del != NULL) {
```
Comment (kommiharibabu) [inline src/btree/bt_debug.c:1642]: Just to point out that this code change is not related to RTS. I am fine to change it in this PR. How about changing the PR title to `RTS and dump` logs?

**Review by kommiharibabu — APPROVED** [review]
LGTM. Thanks.

File: src/btree/bt_debug.c (hunk 1 above)
Comment (etienneptl) [reply src/btree/bt_debug.c:1642]: Sure thing, done!

File: src/txn/txn_rollback_to_stable.c [hunk 2]
```diff
@@ -1231,23 +1230,23 @@ static int
...
+    if (!modified && !__rollback_page_needs_abort(session, ref, rollback_timestamp)) {
         __wt_verbose_multi(session, WT_VERB_RECOVERY_RTS(session), "%p: page skipped", (void *)ref);
         return (0);
     }
 
     WT_STAT_CONN_INCR(session, txn_rts_pages_visited);
     __wt_verbose_multi(session, WT_VERB_RECOVERY_RTS(session),
-      "%p: page rolled back when page is modified: %s", (void *)ref,
-      __wt_page_is_modified(page) ? "true" : "false");
+      "%p: page rolled back when page is modified: %s", (void *)ref, modified ? "true" : "false");
```
Comment (sueloverso) [inline src/txn/txn_rollback_to_stable.c:1249]: I find reading the same text with true or false more difficult than different text. Maybe something like:
"%p: roll back %s page", (void *)ref, modified ? "modified" : "clean" 

Or "clean with <something>"

**Review by sueloverso — APPROVED** [review]
LGTM - I added a couple more suggestions for consideration.

File: src/txn/txn_rollback_to_stable.c [hunk 3]
```diff
@@ -1231,23 +1230,23 @@ static int
...
      * If we have a ref with clean page, find out whether the page has any modifications that are
      * newer than the given timestamp. As eviction writes the newest version to page, even a clean
      * page may also contain modifications that need rollback.
      */
     page = ref->page;
-    if (!__wt_page_is_modified(page) &&
-      !__rollback_page_needs_abort(session, ref, rollback_timestamp)) {
+    modified = __wt_page_is_modified(page);
+    if (!modified && !__rollback_page_needs_abort(session, ref, rollback_timestamp)) {
         __wt_verbose_multi(session, WT_VERB_RECOVERY_RTS(session), "%p: page skipped", (void *)ref);
```
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c:1243]: Yes, `stable` is correct.

[the remark this answers is not in the document]
Comment (etienneptl) [reply src/txn/txn_rollback_to_stable.c:1243]: Should we increment `txn_rts_tree_walk_skip_pages` here @kommiharibabu ?

[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c:1243]: The tree walk logic already found this page because this page is in memory. Later we are verifying whether the page is modified or not to decide whether the page has unstable updates or not after taking the hazard pointer on the page. So I feel this stat may not be the correct one, I don't think we have any other stat to indicate the page skip.

**Comment by etienneptl** [comment]
@sueloverso @kommiharibabu, I have reworked a few RTS logs and pushed those changes using different commits to you can focus on one at a time. If you think one or several or all changes are not worth it, I can revert them easily, just let me know.
I tried to make the logs _easier_ to read by printing out the information used by each check that decides whether something should rolled back or not.

Logs are following this format:
```
Rolling back because a (<value_a>) == b (<value_b>): true
```

I think it has multiple benefits:

- We can see if a value seems off
- We can see what condition was valid at the time of RTS

However, it makes RTS a tiny bit more verbose.

File: src/txn/txn_rollback_to_stable.c [hunk 4]
```diff
@@ -114,14 +116,18 @@ __rollback_abort_update(WT_SESSION_IMPL *session, WT_ITEM *key, WT_UPDATE *first
...
+        if (!txn_id_visible || rollback_timestamp < upd->durable_ts ||
+          upd->prepare_state == WT_PREPARE_INPROGRESS) {
             __wt_verbose_multi(session, WT_VERB_RECOVERY_RTS(session),
               "rollback to stable update aborted with txnid: %" PRIu64
-              " durable timestamp: %s and stable timestamp: %s, prepared: %s",
-              upd->txnid, __wt_timestamp_to_string(upd->durable_ts, ts_string[0]),
+              ", txnid not visible: %s, stable timestamp (%s) < durable timestamp (%s): %s, "
+              "prepared: %s",
+              upd->txnid, !txn_id_visible ? "true" : "false",
+              __wt_timestamp_to_string(upd->durable_ts, ts_string[0]),
```
Comment (sueloverso) [inline src/txn/txn_rollback_to_stable.c:127]: I think this line needs to be below the rollback_timestamp line. The print format has 'stable timestamp' first then 'durable timestamp' and these calls have `upd->durable_ts` then `rollback_timestamp`.

File: src/txn/txn_rollback_to_stable.c [hunk 5]
```diff
@@ -114,14 +116,18 @@ __rollback_abort_update(WT_SESSION_IMPL *session, WT_ITEM *key, WT_UPDATE *first
...
               __wt_timestamp_to_string(rollback_timestamp, ts_string[1]),
-              rollback_timestamp < upd->durable_ts ? "false" : "true");
+              rollback_timestamp < upd->durable_ts ? "true" : "false",
+              upd->prepare_state == WT_PREPARE_INPROGRESS ? "true" : "false");
```
Comment (sueloverso) [inline src/txn/txn_rollback_to_stable.c:130]: Would it be useful to print out the prepared state? If it is not INPROGRESS, it might be helpful to know what the state is.

File: src/txn/txn_rollback_to_stable.c [hunk 6]
```diff
@@ -718,11 +724,16 @@ __rollback_abort_ondisk_kv(WT_SESSION_IMPL *session, WT_REF *ref, WT_ROW *rip, u
...
-          "prepared: %s, stable timestamp: %s and txnid : %" PRIu64,
-          __wt_timestamp_to_string(vpack->tw.durable_start_ts, ts_string[0]),
-          __wt_timestamp_to_string(vpack->tw.start_ts, ts_string[1]), prepared ? "true" : "false",
-          __wt_timestamp_to_string(rollback_timestamp, ts_string[2]), vpack->tw.start_txn);
+          "on-disk update with commit timestamp %s aborted because start durable timestamp (%s) > "
+          "stable timestamp "
+          "(%s): %s, or txnid (%" PRIu64
+          ") is not visible: %s, or tw has not stop and is prepared: %s",
+          __wt_timestamp_to_string(vpack->tw.start_ts, ts_string[0]),
+          __wt_timestamp_to_string(vpack->tw.durable_start_ts, ts_string[1]),
```
Comment (sueloverso) [inline src/txn/txn_rollback_to_stable.c:732]: Again, these two `wt_timestamp_to_string` lines seem out of order. The format says durable > stable, but these lines are start_ts then durable_start_ts. Oh  wait, the `start_ts` goes with the `commit timestamp`.

I suggest breaking this one verbose statement into several. The format has so many parts and the various conditions to apply to the format make this very hard to read. Changing it correctly could be challenging.

File: src/txn/txn_rollback_to_stable.c [hunk 7]
```diff
@@ -718,11 +724,16 @@ __rollback_abort_ondisk_kv(WT_SESSION_IMPL *session, WT_REF *ref, WT_ROW *rip, u
...
+          __wt_timestamp_to_string(rollback_timestamp, ts_string[2]),
+          vpack->tw.durable_start_ts > rollback_timestamp ? "true" : "false", vpack->tw.start_txn,
+          !__rollback_txn_visible_id(session, vpack->tw.start_txn) ? "true" : "false",
+          !WT_TIME_WINDOW_HAS_STOP(&vpack->tw) && prepared ? "true" : "false");
```
Comment (sueloverso) [inline src/txn/txn_rollback_to_stable.c:736]: I personally find these complex messages with various true/false hard to read and reason about. But if this is what others find useful then that's fine.

File: src/txn/txn_rollback_to_stable.c [hunk 8]
```diff
@@ -1231,23 +1242,24 @@ static int
...
+          "%p: unmodified stable page skipped", (void *)ref);
         return (0);
     }
 
     WT_STAT_CONN_INCR(session, txn_rts_pages_visited);
-    __wt_verbose_multi(session, WT_VERB_RECOVERY_RTS(session),
-      "%p: page rolled back when page is modified: %s", (void *)ref,
-      __wt_page_is_modified(page) ? "true" : "false");
+    __wt_verbose_multi(session, WT_VERB_RECOVERY_RTS(session), "%p: roll back %s page", (void *)ref,
+      modified ? "modified" : "clean");
```
Comment (sueloverso) [inline src/txn/txn_rollback_to_stable.c:1262]: This is the sort of message I find more helpful. Rather than "modified: false" seeing "modified" or "clean" means I don't have to apply the boolean to what is written.

File: src/txn/txn_rollback_to_stable.c [hunk 9]
```diff
@@ -1782,13 +1802,19 @@ __rollback_to_stable_btree_apply(
...
+          S2BT(session)->modified ? "true" : "false",
           __wt_timestamp_to_string(max_durable_ts, ts_string[0]),
-          S2BT(session)->modified ? "true" : "false", prepared_updates ? "true" : "false",
-          !durable_ts_found ? "true" : "false", rollback_txnid,
+          __wt_timestamp_to_string(rollback_timestamp, ts_string[1]),
+          max_durable_ts > rollback_timestamp ? "true" : "false",
+          prepared_updates ? "true" : "false", !durable_ts_found ? "true" : "false", rollback_txnid,
+          S2C(session)->recovery_ckpt_snap_min,
           has_txn_updates_gt_than_ckpt_snap ? "true" : "false");
+
```
Comment (sueloverso) [inline src/txn/txn_rollback_to_stable.c:1817]: Errant new line?

File: src/txn/txn_rollback_to_stable.c [hunk 10]
```diff
@@ -1782,13 +1802,19 @@ __rollback_to_stable_btree_apply(
...
+          ") > recovery checkpoint snap min (%" PRIu64 "): %s",
+          S2BT(session)->modified ? "true" : "false",
           __wt_timestamp_to_string(max_durable_ts, ts_string[0]),
-          S2BT(session)->modified ? "true" : "false", prepared_updates ? "true" : "false",
-          !durable_ts_found ? "true" : "false", rollback_txnid,
+          __wt_timestamp_to_string(rollback_timestamp, ts_string[1]),
+          max_durable_ts > rollback_timestamp ? "true" : "false",
+          prepared_updates ? "true" : "false", !durable_ts_found ? "true" : "false", rollback_txnid,
+          S2C(session)->recovery_ckpt_snap_min,
           has_txn_updates_gt_than_ckpt_snap ? "true" : "false");
```
Comment (sueloverso) [inline src/txn/txn_rollback_to_stable.c:1816]: This is another one that is a style that's hard for me to parse and understand when looking at verbose output. It is also again so large and complex it may be more helpful to split it up. I also recognize that by splitting it up that it could be separated in the output stream if other threads get in between multiple lines.

**Comment by sueloverso** [comment]
I think there might be a reversed set of timestamps printed out in one of the statements. I personally find this true/false message style (both existing and these) challenging to parse and understand. But these changes are generally an improvement. I'd like one more look (if you decide to split some long, complex statements).

File: src/txn/txn_rollback_to_stable.c (hunk 9 above)
Comment (etienneptl) [reply src/txn/txn_rollback_to_stable.c:1817]: I like separating stuff/sections of code to improve (my) readability, let me know if I should revert.

File: src/txn/txn_rollback_to_stable.c (hunk 8 above)
Comment (etienneptl) [reply src/txn/txn_rollback_to_stable.c:1262]: I do prefer those too, easier to read. I am just not sure how to apply this format where there are complex checks on whether we decide to roll back or not.

File: src/txn/txn_rollback_to_stable.c (hunk 5 above)
Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c:130]: During RTS, we have either IN PROGRESS or RESOLVED states that are possible. The intermediate LOCKED state is not possible, so whenever we are going to remove the update, it is in the state of IN PROGRESS only.

File: src/txn/txn_rollback_to_stable.c [hunk 11]
```diff
@@ -718,11 +724,16 @@ __rollback_abort_ondisk_kv(WT_SESSION_IMPL *session, WT_REF *ref, WT_ROW *rip, u
       !__rollback_txn_visible_id(session, vpack->tw.start_txn) ||
       (!WT_TIME_WINDOW_HAS_STOP(&vpack->tw) && prepared)) {
         __wt_verbose_multi(session, WT_VERB_RECOVERY_RTS(session),
-          "on-disk update aborted with start durable timestamp: %s, commit timestamp: %s, "
-          "prepared: %s, stable timestamp: %s and txnid : %" PRIu64,
-          __wt_timestamp_to_string(vpack->tw.durable_start_ts, ts_string[0]),
-          __wt_timestamp_to_string(vpack->tw.start_ts, ts_string[1]), prepared ? "true" : "false",
-          __wt_timestamp_to_string(rollback_timestamp, ts_string[2]), vpack->tw.start_txn);
+          "on-disk update with commit timestamp %s aborted because start durable timestamp (%s) > "
```
Comment (kommiharibabu) [inline src/txn/txn_rollback_to_stable.c:727]: We have a function that prints the entire time point `__wt_time_point_to_string`. How do you feel printing the time point at one place and the comparison results later?

File: src/txn/txn_rollback_to_stable.c (hunk 5 above)
Comment (etienneptl) [reply src/txn/txn_rollback_to_stable.c:130]: I guess I will leave it as is then if there are only two choices.

File: src/txn/txn_rollback_to_stable.c (hunk 11 above)
Comment (etienneptl) [reply src/txn/txn_rollback_to_stable.c:727]: @kommiharibabu, do you mean something like this:
```
         __wt_verbose_multi(session, WT_VERB_RECOVERY_RTS(session),
-          "on-disk update with commit timestamp %s aborted because start durable timestamp (%s) > "
-          "stable timestamp "
-          "(%s): %s, or txnid (%" PRIu64
-          ") is not visible: %s, or tw has not stop and is prepared: %s",
-          __wt_timestamp_to_string(vpack->tw.start_ts, ts_string[0]),
-          __wt_timestamp_to_string(vpack->tw.durable_start_ts, ts_string[1]),
-          __wt_timestamp_to_string(rollback_timestamp, ts_string[2]),
-          vpack->tw.durable_start_ts > rollback_timestamp ? "true" : "false", vpack->tw.start_txn,
+          "on-disk update aborted with time window %s. Start durable timestamp > stable timestamp: "
+          "%s, or txnid is not visible: %s, or tw has not stop and is prepared: %s",
+          __wt_time_point_to_string(
+            vpack->tw.start_ts, vpack->tw.durable_start_ts, vpack->tw.start_txn, time_string),
+          vpack->tw.durable_start_ts > rollback_timestamp ? "true" : "false",
           !__rollback_txn_visible_id(session, vpack->tw.start_txn) ? "true" : "false",
           !WT_TIME_WINDOW_HAS_STOP(&vpack->tw) && prepared ? "true" : "false");
```

Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c:727]: Yes. This way the time window of an on-disk entry or update is printed at the start of the message and the result decision is later. I feel It avoids mistakes.

Comment (etienneptl) [reply src/txn/txn_rollback_to_stable.c]: Thanks, pushed as part of 7ce9cc1d825443b90fe45ed3070265a8458ecf8c

**Review by kommiharibabu — APPROVED** [review]
LGTM

**Comment by etienneptl** [comment]
evergreen merge
